### DISCUSSION TITLE
how to set schema for array of objects

### DISCUSSION THREAD

**Discussion by dtau00** [appy-one/acebase#127] [Q&A] [answered]
Here's a simple example:

await db.schema.set('chats/$chatid',{
  id: 'string',
  messages:'{message: "string"}[]'  // im assuming this is the correct syntax
})

so now i want to add a message object to the messages array, via proxy

chat.messages.push({
  message: 'hello'
})

and im getting the error:

"Failed to update server value for "...", rolling back cache to previous value. Error: Error: Schema validation failed: every array value of path "..." must match one of the specified types"

I've tried various different syntaxes and whatnot, this seems to be the most correct.  but cant seem to get it to work.

**Comment by appy-one** [comment, marked as answer]
Although I would recommend not using arrays of objects (use object collections instead!), the error is in your schema definition.

By setting `messages` to `{message: "string"}[]` you are enforcing every `message` to have the value `"string"` instead of being of type `string`. Remove the quotes around string and you are good to go:

```js
await db.schema.set('chats/$chatid',{
  id: 'string',
  messages:'{ message: string }[]' // No quotes around `string` type definition
})
```

**Reply by dtau00** [reply]
cool, easy enough

**Comment by appy-one** [comment]
Sidenote about why not to use arrays: see [Using arrays](https://github.com/appy-one/acebase#using-arrays)

Additional info about using arrays with a live data proxy: any change to any item in the array will cause the entire array to be rewritten because of reasons described in above link. If you use object collections instead, only the actual changes are written and synced.

To change your code to use object collections instead, change your schema to:
```js
await db.schema.set('chats/$chatid',{
  id: 'string',
  messages: { 
    '*': { message: 'string' }
  }
});
```

Of course you will then have to set the initial value of a chat.messages to `{}` instead of `[]`

Your `.push` on the proxy `messages` will also work:
```js
chat.messages.push({ message: 'Hello' }); // stored in eg "[chat path]/messages/l5f4ll26000109mh76lma34d"
```

**Reply by dtau00** [reply]
oh is that right? i thought the entire array wont be rewritten as long as you're pushing onto the array.  i was using arrays for chat messages since they should be sequential, and i wont have to keep reordering the messages.

**Reply by appy-one** [reply]
The generated keys in object collections (cuids) are time-sortable: a key generated now > a key generated before. There's no need to order the messages in the collection manually because they will automatically be in the right order
